keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,925 @@
+
+*Converted from this [original proposal doc](http://bit.ly/k8s-mc-svc-api-proposal).*
+
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this KEP.  Describe why the change is important and the benefits to users.  The
+motivation section can optionally provide links to [experience reports][] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+There are [many
+reasons](http://bit.ly/k8s-multicluster-conversation-starter-doc) why a K8s user
+may want to split their deployments across multiple clusters, but still retain
+mutual dependencies between workloads running in those clusters. Today the
+cluster is a hard boundary, and a service is opaque to a remote K8s consumer
+that would otherwise be able to make use of metadata (e.g. endpoint toplogy) to
+better direct traffic. To support failover or temporarily during migration,
+users may want to consume services spread across clusters, but today that
+requires non-trivial bespoke solutions.
+
+The Multi-Cluster Services API aims to fix these problems.
+
+### Goals
+
+<!--
+List the specific goals of the KEP.  What is it trying to achieve?  How will we
+know that this has succeeded?
+-->
+- Define a minimal API to support service discovery and consumption across clusters.
+  - Consume a service in another cluster.
+  - Consume a service deployed in multiple clusters as a single service.
+- When a service is consumed from another cluster its behavior should be
+  predictable and consistent with how it would be consumed within its own
+  cluster.
+- Create building blocks for multi-cluster tooling.
+- Support multiple implementations.
+- Leave room for future extension and new use cases.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Define specific implementation details beyond general API behavior.
+- Change behavior of single cluster services in any way.
+- Define what NetworkPolicy means for multi-cluster services.
+- Solve mechanics of multi-cluster service orchestration.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+#### Terminology
+
+- **supercluster** - A placeholder name for a group of clusters with a high
+  degree of mutual trust and shared ownership that share services amongst
+  themselves. Membership in a supercluster is symmetric and transitive. The set
+  of member clusters are mutually aware, and agree about their collective
+  association.
+- **mcsd-controller** - A controller that syncs services across clusters and
```

Comment:
Since `mcsd-controller` is such a key component of the solution, it would be good to have a common section that lists all expectations/assumptions any implementation of such a controller would meet.

Reply:
We also need not to be prescriptive of an specific implementation. We have moved subm.Lighthouse for  to be de-centralized now.

Reply:
Yeah, it's purposefully vague right now. The hope is that this API and behavioral description can be enough to cover what's necessary to start to build common tooling over multi-cluster services, but leave tons of room for flexibility in implementation. The way I see it is that a service owner should know that creating a `ServiceExport` exposes a service, and a consumer should know that an imported service can be consumed via the `*.supercluster.local` name, the VIP, or the `ServiceImport`, but everything in between is up to the implementation

Reply:
I will make this even more vague - there needs to be some controller responsible for this. It does not need to be a single instance, or single process. For the purposes of this API, a person manually copying things around would fit the desciription

Reply:
Maybe coz am about to start working on this, so would've liked it explicitly called out. I'll figure out eventually over time, so not that crucial.